src/wrappers/hapi.js

```diff
@@ -0,0 +1,115 @@
+/* eslint-disable prefer-rest-params */
+/**
+ * @fileoverview Handlers for Hapi instrumentation
+ */
+
+const shimmer = require('shimmer');
+const tryRequire = require('../try_require.js');
+const utils = require('../utils.js');
+const tracer = require('../tracer.js');
+const traceContext = require('../trace_context.js');
+const eventInterface = require('../event.js');
+const hapiRunner = require('../runners/hapi.js');
+
+const Hapi = tryRequire('hapi');
+
+
+/**
+ * Hapi requests middleware
+ * @param {Object} request The Hapi's request data
+ * @param {Object} h The Hapi's response data
+ * @param {Function} originalHandler function for the Hapi's route
+ * @return {Object} response
+ */
+function hapiMiddleware(request, h, originalHandler) {
+    // Initialize tracer
+    tracer.restart();
+
+    let hapiEvent;
+    const startTime = Date.now();
+    try {
+        hapiEvent = hapiRunner.createRunner(request, startTime);
+        tracer.addRunner(hapiEvent);
+    } catch (err) {
+        utils.debugLog(err);
+        return originalHandler(request, h);
+    }
+
+    // Inject trace functions
+    const { label, setError } = tracer;
+    request.epsagon = {
+        label,
+        setError,
+    };
+
+    // Run the request, activate the context
+    const response = originalHandler(request, h);
+
+    // Handle response
+    response.then(() => {
+        try {
+            hapiRunner.finishRunner(hapiEvent, request, response, startTime);
+        } catch (err) {
+            tracer.addException(err);
+        }
+        tracer.sendTrace(() => {});
+    }).catch((err) => {
```

Comment:
The code duplication here is just screaming :(

Reply:
👍